api/prometheus/v1/api.go

```diff
@@ -612,6 +613,7 @@ type RecordingRule struct {
 	LastError      string         `json:"lastError,omitempty"`
 	EvaluationTime float64        `json:"evaluationTime"`
 	LastEvaluation time.Time      `json:"lastEvaluation"`
+	Type           string         `json:"type"`
```

Comment:
If we look on `Rules` type commentary you see the canonical way to identify if it's recording or alerting is through a type system. This means we don't need this type field here.

From the error you are getting, it sounds like you have problem with something marshalling this type back to JSON and then something unmarshalling (decoding) and expecting `type` field? In this case:

1. For sure we should have regression test and agree it's a valid use case. Can you confirm that is the case? Simple call of API that uses unmarshalling should just work without `Type` in this code -- unless I am wrong, but reproduction test will verify (: 
2.  To solve it, we could create `MarshalJSON` function that adds that field just for marshalling, WDYT? Alternatively we could keep this PR, but perhaps change the commentary on `Rules` slice? Otherwise we will have multiple ways of doing same thing.

To sum up, regression test that breaks on old code would be amazing, in minimum, WDYT?

Reply:
So in our use case we simply call the Rules endpoint using the prometheus API here https://github.com/prometheus/client_golang/blob/main/api/prometheus/v1/api.go#L1202. In doing so it calls the ruler endpoint and returns us a RulesResult struct which we just use as a data field in response json. But it seems like when internally in api.go we are trying to create unmarshall RuleResult struct and while doing so unmarshall alertingRule and recordingRule it expects the struct to have the type field to unmarshal the json successfully which it gets from the ruler endpoint.

Reply:
> it calls the ruler endpoint and returns us a RulesResult struct which we just use as a data field in response json. But it seems like when internally in api.go we are trying to create unmarshall RuleResult struct and while doing so unmarshall alertingRule and recordingRule it expects the struct to have the type field to unmarshal the json successfully which it gets from the ruler endpoint.

Cool. So literally Prometheus server is targeted, right? Then Prometheus server should give us `type` field and our code should work without this PR. In fact `alertingRule.Type = string(RuleTypeAlerting)` line does nothing given you added `type` field, but even without type field unmarshal should work because of anonymous struct with type we unmarshal first. 

BUT, if you notice a problem there might be one,  so the best is to create a very small unit test that e.g. mock Prometheus API returning recording and alerting with type specified. Then without your change it should fail, right?

Reply:
We don't have enough tests so it would be amazing anyway 🤗  - if not done I can try do it, but it will take time (busy).